src/livekit/MediaDevicesContext.tsx

```diff
@@ -249,6 +253,144 @@ export const MediaDevicesProvider: FC<Props> = ({ children }) => {
     [setNumCallersUsingNames],
   );
 
+  return {
+    audioInput,
+    videoInput,
+    startUsingDeviceNames,
+    stopUsingDeviceNames,
+    usingNames,
+  };
+}
+
+interface Props {
+  children: JSX.Element;
+}
+
+export const MediaDevicesProvider: FC<Props> = ({ children }) => {
+  const {
+    audioInput,
+    videoInput,
+    startUsingDeviceNames,
+    stopUsingDeviceNames,
+    usingNames,
+  } = useInputDevices();
+
+  const audioOutput = useMediaDevice(
+    "audiooutput",
+    audioOutputSetting,
+    usingNames,
+  );
+
+  const context: MediaDevices = useMemo(
+    () => ({
+      audioInput,
+      audioOutput,
+      videoInput,
+      startUsingDeviceNames,
+      stopUsingDeviceNames,
+    }),
+    [
+      audioInput,
+      audioOutput,
+      videoInput,
+      startUsingDeviceNames,
+      stopUsingDeviceNames,
+    ],
+  );
+
+  return (
+    <MediaDevicesContext.Provider value={context}>
+      {children}
+    </MediaDevicesContext.Provider>
+  );
+};
+
+function useControlledOutput(): MediaDevice {
+  const showEarpiece = useShowEarpiece();
+
+  const available = useObservableEagerState(
+    useObservable(() =>
+      setOutputDevices$.pipe(
+        startWith<OutputDevice[]>([]),
+        map((devices) => {
+          const devicesMap = new Map<string, DeviceLabel>(
+            devices.map(({ id, name }) => [id, { type: "name", name }]),
+          );
+          if (showEarpiece)
+            devicesMap.set(EARPIECE_CONFIG_ID, { type: "earpiece" });
+          return devicesMap;
+        }),
+      ),
+    ),
+  );
+  const earpiceDevice = useObservableEagerState(
+    setOutputDevices$.pipe(
+      map((devices) => devices.find((d) => d.forEarpiece)),
+    ),
+  );
+
+  const [preferredId, setPreferredId] = useSetting(audioOutputSetting);
+
+  const selectedId = useMemo(() => {
+    if (available.size) {
+      // If the preferred device is available, use it. Or if every available
+      // device ID is falsy, the browser is probably just being paranoid about
+      // fingerprinting and we should still try using the preferred device.
+      // Worst case it is not available and the browser will gracefully fall
+      // back to some other device for us when requesting the media stream.
+      // Otherwise, select the first available device.
+      return (preferredId !== undefined && available.has(preferredId)) ||
+        (available.size === 1 && available.has(""))
+        ? preferredId
+        : available.keys().next().value;
+    }
+    return undefined;
+  }, [available, preferredId]);
+
+  useEffect(() => {
+    if (selectedId === EARPIECE_CONFIG_ID)
+      if (selectedId !== undefined)
+        window.controls.onOutputDeviceSelect?.(selectedId);
+  }, [selectedId]);
+
+  const [asEarpice, setAsEarpiece] = useState(false);
+
+  const select = useCallback(
+    (id: string) => {
+      if (id === EARPIECE_CONFIG_ID) {
+        setAsEarpiece(true);
+        if (earpiceDevice) setPreferredId(earpiceDevice.id);
```

Comment:
I feel like we should log a warning or something if there is no `earpieceDevice`

Reply:
With the recent developments in how we need to solve it on ios we actually need the case that there is no earpiece device.

Reply:
If bt headphones are selected the user first has to use the native picker to go back to speaker and then the earpice can be selected.
Modeling it with earpice being available in the list or not is the most elegant we can do I think .